xwiki-commons-core/xwiki-commons-configuration/xwiki-commons-configuration-default/src/main/java/org/xwiki/configuration/internal/SystemEnvConfigurationSource.java

```diff
@@ -0,0 +1,218 @@
+        String encodedCompletePrefix = PREFIX + encodedPrefix;
+
+        return getenv().keySet().stream().filter(k -> isEnvKey(k, encodedCompletePrefix))
+            .map(k -> fromEnvKey(k, prefix, encodedPrefix)).toList();
+    }
+
+    @Override
+    public boolean containsKey(String key)
+    {
+        if (key == null) {
+            return false;
+        }
+
+        return getenv().containsKey(toEnvKey(key));
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return isEmpty("");
+    }
+
+    @Override
+    public boolean isEmpty(String prefix)
+    {
+        Map<String, String> env = getenv();
+
+        if (env.isEmpty()) {
+            return true;
+        }
+
+        String encodedCompletePrefix = toEnvKey(prefix);
+
+        return env.keySet().stream().noneMatch(k -> isEnvKey(k, encodedCompletePrefix));
+    }
+
+    private String encode(String property)
+    {
+        if (property.isEmpty()) {
+            return property;
+        }
+
+        StringBuilder builder = new StringBuilder(property.length() * 3);
+
+        encode(property, builder);
+
+        return builder.toString().toUpperCase(Locale.ROOT);
+    }
+
+    private void encode(String property, StringBuilder builder)
+    {
+        for (int i = 0; i < property.length(); ++i) {
+            char c = property.charAt(i);
+
+            boolean encode = false;
+
+            if (i == 0) {
+                // Environment variable keys cannot start with a digit
+                if (Character.isDigit(c)) {
+                    encode = true;
+                }
+            } else {
+                // Environment variable keys are only allowed to contain underscore or alphanumeric characters
+                if (c != '_' && !Character.isAlphabetic(c) && !Character.isDigit(c)) {
+                    encode = true;
+                }
+            }
+
+            if (encode) {
+                encode(c, builder);
+            } else {
+                builder.append(c);
+            }
+        }
+    }
+
+    private String toEnvKey(String property)
+    {
+        if (property.isEmpty()) {
+            return PREFIX;
+        }
+
+        StringBuilder builder = new StringBuilder(property.length() * 3);
+
+        builder.append(PREFIX);
+
+        encode(property, builder);
+
+        return builder.toString().toUpperCase(Locale.ROOT);
+    }
+
+    /**
+     * @param c the character to encode
+     * @param builder the builder to append the encoded character to
+     */
+    private void encode(char c, StringBuilder builder)
+    {
+        builder.append('_');
+
+        switch (c) {
+            case '.', ':', '-':
+                break;
+            default:
+                byte[] ba = String.valueOf(c).getBytes(StandardCharsets.UTF_8);
+
+                for (int j = 0; j < ba.length; j++) {
+                    char ch = Character.forDigit((ba[j] >> 4) & 0xF, 16);
+                    // Make it upper case
```

Comment:
Well, this comment is not really super useful but I'd have liked a comment on the instruction above :)

Reply:
It's actually mostly a copy/paste of URLEncoder. I will check if it's possible to use `org.apache.commons.codec.net.PercentCodec` without the % instead of hardcoded the encoding code.